File: Sorting/merge_sort.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

// Time Complexity: O(n log n) in all cases (worst, average, best)
// Space Complexity: O(n) due to the use of temporary vectors

void merge(vector<int>& arr, int low, int mid, int high) {
    vector<int> temp;
    int left = low;
    int right = mid + 1;

    while (left <= mid && right <= high) {
        if (arr[left] < arr[right]) temp.push_back(arr[left++]);
        else temp.push_back(arr[right++]);
    }

    while (left <= mid) temp.push_back(arr[left++]);
    while (right <= high) temp.push_back(arr[right++]);

    for (int i = low; i <= high; i++) {
        arr[i] = temp[i - low]; 
    }
}

void mergesort(vector<int>& arr, int low, int high) {
    if (low >= high) return;
    int mid = (low + high) / 2;
    mergesort(arr, low, mid);
    mergesort(arr, mid + 1, high);
    merge(arr, low, mid, high);
}

int main() {
    vector<int> arr = {64, 25, 12, 22, 11};
    mergesort(arr, 0, arr.size() - 1);
    cout << "Sorted array: ";
    for (int num : arr) {
        cout << num << " ";
    }
    return 0;
}

File: java/Sorting/MergeSort.java
package Sorting;

import java.util.*;

// Merge Sort implementation in Java
public class MergeSort {

    /*
     * Time Complexity Intuition:
     * --------------------------
     * - The array is recursively divided into two halves → log₂(n) levels.
     * - At each level, merging takes O(n) time.
     *
     * Total Time Complexity:
     * Worst Case  : O(n log n)
     * Average Case: O(n log n)
     * Best Case   : O(n log n)
     *
     * Space Complexity Intuition:
     * ---------------------------
     * - A temporary array is used during each merge step.
     * - Extra space proportional to the size of the array is required.
     *
     * Space Complexity: O(n)
     */

    // Function to merge two sorted halves of the array
    static void merge(int[] arr, int low, int mid, int high) {

        // Temporary list to store merged elements
        ArrayList<Integer> temp = new ArrayList<>();

        int left = low;       // Pointer for left subarray
        int right = mid + 1;  // Pointer for right subarray

        // Merge elements from both halves in sorted order
        while (left <= mid && right <= high) {
            if (arr[left] < arr[right]) {
                temp.add(arr[left++]);
            } else {
                temp.add(arr[right++]);
            }
        }

        // Copy remaining elements from left half
        while (left <= mid) {
            temp.add(arr[left++]);
        }

        // Copy remaining elements from right half
        while (right <= high) {
            temp.add(arr[right++]);
        }

        // Copy merged elements back to original array
        for (int i = low; i <= high; i++) {
            arr[i] = temp.get(i - low);
        }
    }

    // Recursive merge sort function
    static void mergeSort(int[] arr, int low, int high) {

        // Base case: single element is already sorted
        if (low >= high) return;

        int mid = (low + high) / 2;

        // Sort left half
        mergeSort(arr, low, mid);

        // Sort right half
        mergeSort(arr, mid + 1, high);

        // Merge both halves
        merge(arr, low, mid, high);
    }

    public static void main(String[] args) {
        int[] arr = {64, 25, 12, 22, 11};

        mergeSort(arr, 0, arr.length - 1);

        System.out.print("Sorted array: ");
        for (int num : arr) {
            System.out.print(num + " ");
        }
    }
}
